File: connectors/connector-jira/src/main/scala/com/taskadapter/connector/jira/CustomFieldResolver.scala
package com.taskadapter.connector.jira

import com.atlassian.jira.rest.client.api.domain.{Field, IssueField}
import com.taskadapter.model.{CustomSeqString, CustomString}

object CustomFieldResolver {
  def apply(): CustomFieldResolver = new CustomFieldResolver(Seq())
}

/**
  * @param fullIdForSave e.g. 'customfield_10100'
  * @param typeName sample values: 'array', 'any', 'option', 'progress', 'user', 'number', 'datetype', ...
  * @param itemsTypeIfArray "null" for any [typeName] other than 'array'. for arrays sample values are:
  *                         "string","attachment", "version", "component", "version", "issueLinks", "worklog", ...
  */
case class JiraFieldDefinition(id: Long, fieldName: String, fullIdForSave: String, typeName: String, itemsTypeIfArray: Option[String])

/**
  * Sample custom field loaded from JIRA 7:
  *
  * ```
   Field{id=customfield_10100, name=label_environment, fieldType=CUSTOM, orderable=true, navigable=true, searchable=true,
         schema=FieldSchema{type=array, items=string, system=null,
                            custom=com.atlassian.jira.plugin.system.customfieldtypes:labels, customId=10100}}
  * ```
  */
class CustomFieldResolver(fields: Iterable[Field]) {
  val mapNameToSchema = fields
    .filter(f => f.getSchema != null)
    .map(f => f.getName -> JiraFieldDefinition(f.getSchema.getCustomId, f.getName, f.getId, f.getSchema.getType, Option(f.getSchema.getItems)))
    .toMap

  def getId(name: String): Option[JiraFieldDefinition] = {
    mapNameToSchema.get(name)
  }

  def getField(jiraField: IssueField): Option[com.taskadapter.model.Field[_]] = {
    getId(jiraField.getName) match {
      case Some(definition) => definition.typeName match {
        case "string" | "any" => Some(CustomString(definition.fieldName))
        case "array" => Some(CustomSeqString(definition.fieldName))
        case _ => None
      }
      case None => None
    }

  }
}

File: connectors/connector-jira/src/main/java/com/taskadapter/connector/jira/CustomFieldResolver.java
package com.taskadapter.connector.jira;

import com.atlassian.jira.rest.client.api.domain.Field;
import com.atlassian.jira.rest.client.api.domain.IssueField;
import com.taskadapter.model.CustomListString;
import com.taskadapter.model.CustomString;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

/**
 * Sample custom field loaded from JIRA 7:
 *
 * <pre>
 * Field{id=customfield_10100, name=label_environment, fieldType=CUSTOM, orderable=true, navigable=true, searchable=true,
 * schema=FieldSchema{type=array, items=string, system=null,
 * custom=com.atlassian.jira.plugin.system.customfieldtypes:labels, customId=10100}}
 * </pre>>
 */
public class CustomFieldResolver {

    public static CustomFieldResolver apply() {
        return new CustomFieldResolver(List.of());
    }

    private final Map<String, JiraFieldDefinition> nameToCustomFieldDefinition;

    public CustomFieldResolver(Iterable<Field> fields) {
        nameToCustomFieldDefinition = StreamSupport.stream(fields.spliterator(), false)
                .filter(f -> f.getSchema() != null)
                // only keep custom fields
                .filter(f -> f.getSchema().isCustom())
                .collect(Collectors.toMap(Field::getName,
                        f -> new JiraFieldDefinition(f.getSchema().getCustomId(),
                                f.getName(), f.getId(), f.getSchema().getType(),
                                Optional.ofNullable(f.getSchema().getItems())))
                );
    }

    Optional<JiraFieldDefinition> getId(String name) {
        return Optional.ofNullable(nameToCustomFieldDefinition.get(name));
    }

    Optional<com.taskadapter.model.Field<?>> getField(IssueField jiraField) {
        var fieldId = getId(jiraField.getName());
        if (fieldId.isEmpty()) {
            return Optional.empty();
        }
        JiraFieldDefinition definition = fieldId.get();
        return switch (definition.getTypeName()) {
            case "string", "any" -> Optional.of(new CustomString(definition.getFieldName()));
            case "array" -> Optional.of(new CustomListString(definition.getFieldName()));
            default -> Optional.empty();
        };
    }
}
